Signup is successful only when t&c is accepted

MayaSatishRao opened the pull request:
# Descripción
It is made mandatory for new users to accept the terms and conditions in the signin page. If the checkbox regarding T&C is not checked then an error popup displays the message "Please read and accept the terms and conditions".

Fixes: https://github.com/omegaup/omegaup/issues/7394

# Comentarios
I have created a new error message called privacyPolicyNotAccepted and this is translated to all 3 languages. (I ran ./stuff/lint.sh) 

#Screenshots
![Screenshot 2024-01-13 170227](https://github.com/omegaup/omegaup/assets/67920373/5260ba99-5ad0-4164-87aa-a4fab82f3550)

# Checklist:

- [ ] El código sigue la [guía de
      estilo](https://github.com/omegaup/omegaup/wiki/Coding-guidelines) de
      omegaUp.
- [ ] Se corrieron todas las pruebas y pasaron.
- [ ] Si se está agregando funcionalidad nueva, se agregaron pruebas.
- [ ] Si el cambio es grande (> 200 líneas), hay que intentar partirlo en
      varios pull requests. De preferencia uno para los controladores + phpunit
      y luego otro para la interfaz.

MayaSatishRao commented:
@pabo99 Can you review this pr?

pabo99 commented on frontend/templates/en.lang:
```diff
@@ -981,6 +981,7 @@ passwordResetResetTitle = "Reset password"
 passwordResetTokenWasGeneratedSuccessfully = "Token was generated succesfully"
 pendingRegistrations = "Pending requests"
 pendingRegistrationsShowAll = "Show all requests"
+privacyPolicyNotAccepted = "Please read and accept the site's [Use and Privacy Policy](http://blog.omegaup.com/privacy-policy/), as well as the [Code of Conduct](https://blog.omegaup.com/codigo-de-conducta-en-omegaup/)."
```
I'm afraid the message component (currently created in jQuery and not yet updated to a Vue component) doesn't support markdown.

You have the following options:

- Remove the links in the text. It would be better to emphasize the section where the check button is missing. This link may help you: https://getbootstrap.com/docs/4.0/components/forms/?#custom-styles
- Implement the necessary changes to make the message accept markdown.

I would rather prefer the first option because we don't want to add more code in jQuery. Probably in the future, we need to make the migration to a Vue component.

MayaSatishRao replied:
@pabo99 I'll remove the links in the text and I will emphasize the checkbox like the screenshot attached. Is that fine?
![Screenshot 2024-01-16 152158](https://github.com/omegaup/omegaup/assets/67920373/cbc20ae4-986d-497c-890f-258333bdcb30)

pabo99 commented on frontend/www/js/omegaup/components/login/Signup.vue:
```diff
@@ -129,6 +130,7 @@ export default class Signup extends Vue {
   password: string = '';
   passwordConfirmation: string = '';
   recaptchaResponse: string = '';
+  checked: boolean = false;
```
What do you think about changing the name of this variable to something more specific? I suggest `termsAndPolicies` or something that better expresses its meaning

MayaSatishRao replied:
ok I will change the name from checked to termsAndPolicies as you suggested.

pabo99 reviewed (commented):
Your change probably alters the workflow in a Cypress test, and that's why a test is failing. If you need help fixing the test, don't hesitate to ask us.

heduenas commented:
@MayaSatishRao we have an automated test that verifies that user registration works end to end. Unfortunately, the current setup of the test does not perform the click on the t&c button. Can you please update that test to check the t&c box? https://github.com/omegaup/omegaup/blob/main/cypress/e2e/basic_commands.cy.ts#L35

pabo99 commented:
> @heduenas I actually don't know how to write test cases in Cypress. First, I will make the suggested changes. Can you please help me out with the testing because I have never done it before?

Hi, @MayaSatishRao!

No worries; you don't need to create a new test. Identifying where the existing one is failing will be enough.

In this particular case, your change is impacting the behavior of the test `Should register a user` because users will now receive a warning message when attempting to register a new account if they don't accept the 'Privacy Policy' and the 'Code of Conduct'.

So, in this function:

https://github.com/omegaup/omegaup/blob/ce17c4f43c06cfd4b0bd8051526aa486f26c52f0/cypress/support/pageObjects/loginPage.ts#L30-L38

you need to add the following line:

```typescript
cy.get('[data-signup-accept-policies]').check();
```

before the `data-signup-submit` button is clicked.

Please note that `data-signup-accept-policies` will be a new data attribute that you need to add to the checkbox where users should click to accept the policies and terms.

I hope this helps you resolve the error.

MayaSatishRao commented:
@pabo99 Thanks a lot. I'll surely incorporate that and a make a PR

MayaSatishRao commented:
@pabo99 I have made the changes. I have modified the Cypress test and removed the links in the error message popup. I have added required field to the checkbox so that it will be emphasised.

pabo99 commented on frontend/www/js/omegaup/login/signin.ts:
```diff
@@ -84,11 +85,11 @@ OmegaUp.on('ready', () => {
               return;
             }
-            api.User.create({
+           /* api.User.create({
```
nit:
```suggestion
           api.User.create({
```

pabo99 commented on frontend/www/js/omegaup/login/signin.ts:
```diff
@@ -98,10 +99,10 @@ OmegaUp.on('ready', () => {
                   username,
                   password,
-                  /*isAccountCreation=*/ true,
+                  /*isAccountCreation= true,
```
nit:
```suggestion
                  /*isAccountCreation*/= true,
```

pabo99 commented on frontend/www/js/omegaup/login/signin.ts:
```diff
@@ -98,10 +99,10 @@ OmegaUp.on('ready', () => {
                 );
               })
-              .catch(ui.apiError);
+              .catch(ui.apiError);*/
```
nit:
```suggestion
              .catch(ui.apiError);
```

pabo99 reviewed (commented):
Thank you for applying the suggestions!

Just minor changes, and this PR is okay for me.

MayaSatishRao reviewed (commented):
I have also removed the unnecessary log statement. Any more changes?

pabo99 commented on frontend/www/js/omegaup/login/signin.ts:
```diff
@@ -83,7 +84,10 @@ OmegaUp.on('ready', () => {
               return;
             }
-
+            if (termsAndPolicies != true) {
```
nit:
```suggestion
            if (!termsAndPolicies) {
```

pabo99 reviewed (commented):
Just a minor change.

MayaSatishRao commented:
Done!!

pabo99 reviewed (approved):
LGTM.

Thank you for your help solving this issue!
